src/storage/compression/roaring.cpp

```diff
@@ -0,0 +1,1211 @@
+		//
+		// 'x': bits to set to 0 in the result
+
+		idx_t right_bits = start % ValidityMask::BITS_PER_VALUE;
+		idx_t bits_to_set = ValidityMask::BITS_PER_VALUE - right_bits;
+		idx_t left_bits = 0;
+		if (start + bits_to_set > end) {
+			// Limit the amount of bits to set
+			left_bits = (start + bits_to_set) - end;
+			bits_to_set = end - start;
+		}
+
+		// Prepare the mask
+		validity_t mask = ValidityUncompressed::LOWER_MASKS[right_bits];
+		if (left_bits) {
+			// Mask off the part that we don't want to touch (if the range doesn't fully cover the bits)
+			mask |= ValidityUncompressed::UPPER_MASKS[left_bits];
+		}
+
+		idx_t entry_idx = start / ValidityMask::BITS_PER_VALUE;
+		start += bits_to_set;
+		result_data[entry_idx] &= mask;
+	}
+
+	idx_t remaining_bits = end - start;
+	idx_t full_entries = remaining_bits / ValidityMask::BITS_PER_VALUE;
+	idx_t entry_idx = start / ValidityMask::BITS_PER_VALUE;
+	// Set all the entries that are fully covered by the range to 0
+	for (idx_t i = 0; i < full_entries; i++) {
+		result_data[entry_idx + i] = (validity_t)0;
+	}
+
+	if ((remaining_bits % ValidityMask::BITS_PER_VALUE) != 0) {
+		// The last entry touched by the range is only partially covered
+
+		// +======================================+
+		// |                         |xxxxxxxxxxxx|
+		// +======================================+
+		//
+		// 'x': bits to set to 0 in the result
+
+		idx_t bits_to_set = end % ValidityMask::BITS_PER_VALUE;
+		idx_t left_bits = ValidityMask::BITS_PER_VALUE - bits_to_set;
+		validity_t mask = ValidityUncompressed::UPPER_MASKS[left_bits];
+		idx_t entry_idx = end / ValidityMask::BITS_PER_VALUE;
+		result_data[entry_idx] &= mask;
+	}
+}
+
+struct ContainerMetadata {
+public:
+	ContainerMetadata() {
+	}
+	ContainerMetadata(bool is_inverted, bool is_run, uint16_t value) {
+		data.run_container.is_inverted = is_inverted;
+		if (is_run) {
+			data.run_container.is_run = 1;
+			data.run_container.number_of_runs = value;
+			data.run_container.unused = 0;
+		} else {
+			data.non_run_container.is_run = 0;
+			data.non_run_container.cardinality = value;
+			data.non_run_container.unused = 0;
+		}
+	}
+
+public:
+	bool IsRun() const {
+		return data.run_container.is_run;
+	}
+
+	bool IsUncompressed() const {
+		return !data.non_run_container.is_run && data.non_run_container.cardinality == MAX_ARRAY_IDX + 1;
+	}
+
+	bool IsInverted() const {
+		return data.run_container.is_inverted;
+	}
+
+	uint16_t NumberOfRuns() const {
+		return data.run_container.number_of_runs;
+	}
+
+	uint16_t Cardinality() const {
+		return data.non_run_container.cardinality;
+	}
+
+public:
+	union {
+		struct {
+			uint16_t unused : 1;          //! Currently unused bits
+			uint16_t is_inverted : 1;     //! Indicate if this maps nulls or non-nulls
+			uint16_t is_run : 1;          //! Indicate if this is a run container
+			uint16_t unused2 : 2;         //! Currently unused bits
+			uint16_t number_of_runs : 11; //! The number of runs
```

Comment:
IIRC bit-fields are not portable to MSVC

Reply:
Or rather, the actual memory layout of the struct resulting from bitfields is not the same across compilers, but maybe you don't memcpy/serialize these anyway in which case its probably fine.

Reply:
Err, I think I do
This is what we write as metadata for each container

I guess I'll rework this, thanks for the heads up 👍

Reply:
@Maxxen I'm curious though, my union is 2 bytes, I don't see how they can manage to screw up bitfields across compilers when the bitfields shouldn't even be considered at the serialization layer.
I would understand endianness messing this up, but that's not what we're talking about right?

Reply:
The bitfields refer to specific bits in the `uint16_t` - if MSVC puts `number_of_runs` as the highest bits, and clang puts `number_of_runs` as the lowest bits, then the physical layout of these structs differ. As a result a database file generated on Windows cannot be read (or will be read incorrectly) on MacOS.

I think this is a great point and something we don't really test in the CI (but probably should). Checked-in database files might catch this - but we don't have any for the roaring bitmap compression. Perhaps we should add one.

Reply:
Hmm fair enough, I'm reworking the metadata now anyways
now it'll be:

[2 bits: `is_run` and `is_inverted`] * `container_count`
[6 bits: `number_of_runs`] * `run_container_count`
[8 bits: `cardinality`] * `array_or_bitset_container_count`

first two of those are bitpacked